File: internal/app/preprocess.go
package app

import (
	"regexp"
	"strings"
)

// boldStarRE / boldUnderscoreRE match a bold span whose inner edges may carry
// stray spaces (e.g. "** word **"). The capture is the whole interior,
// including any surrounding spaces, so tightenEmphasis can trim it. The inner
// run excludes the delimiter char so two adjacent bold spans don't merge and a
// "***" run is never mistaken for an empty bold span.
var (
	boldStarRE       = regexp.MustCompile(`\*\*([ \t]*[^*]*?[ \t]*)\*\*`)
	boldUnderscoreRE = regexp.MustCompile(`__([ \t]*[^_]*?[ \t]*)__`)
)

// tightenBoldDelimiters removes spaces that sit directly inside bold delimiters
// so "** parola **" renders as bold. CommonMark requires the delimiter to hug
// the text ("**parola**"); glance is more forgiving. Only bold (** ** and
// __ __) is touched — single-star/underscore italic is left alone to avoid
// clashing with bullet lists. Fenced code blocks and inline code spans are
// preserved verbatim.
func tightenBoldDelimiters(src string) string {
	lines := strings.Split(src, "\n")
	inFence := false
	for idx, line := range lines {
		trimmed := strings.TrimLeft(line, " \t")
		if strings.HasPrefix(trimmed, "```") || strings.HasPrefix(trimmed, "~~~") {
			inFence = !inFence
			continue
		}
		if inFence {
			continue
		}
		lines[idx] = tightenBoldOutsideCode(line)
	}
	return strings.Join(lines, "\n")
}

// tightenBoldOutsideCode applies the bold tightening only to the parts of a
// line that are not inside an inline code span (`...`). Backtick runs are
// matched by equal length, mirroring CommonMark's code-span rule.
func tightenBoldOutsideCode(line string) string {
	var b strings.Builder
	i, n := 0, len(line)
	for i < n {
		if line[i] == '`' {
			j := i
			for j < n && line[j] == '`' {
				j++
			}
			runLen := j - i
			if close := indexBacktickRun(line, j, runLen); close >= 0 {
				b.WriteString(line[i : close+runLen]) // code span, verbatim
				i = close + runLen
				continue
			}
			b.WriteString(line[i:j]) // unterminated run: literal backticks
			i = j
			continue
		}
		start := i
		for i < n && line[i] != '`' {
			i++
		}
		seg := line[start:i]
		seg = tightenEmphasis(seg, "**", boldStarRE)
		seg = tightenEmphasis(seg, "__", boldUnderscoreRE)
		b.WriteString(seg)
	}
	return b.String()
}

// indexBacktickRun returns the start index of the next run of exactly runLen
// backticks at or after from, or -1 if there is none.
func indexBacktickRun(s string, from, runLen int) int {
	for i := from; i < len(s); {
		if s[i] == '`' {
			j := i
			for j < len(s) && s[j] == '`' {
				j++
			}
			if j-i == runLen {
				return i
			}
			i = j
		} else {
			i++
		}
	}
	return -1
}

// tightenEmphasis trims spaces directly inside each delim..delim span matched by
// re. A span that is empty after trimming (e.g. "** **") is left untouched.
func tightenEmphasis(s, delim string, re *regexp.Regexp) string {
	return re.ReplaceAllStringFunc(s, func(m string) string {
		inner := m[len(delim) : len(m)-len(delim)]
		trimmed := strings.TrimSpace(inner)
		if trimmed == "" || trimmed == inner {
			return m
		}
		return delim + trimmed + delim
	})
}

// blankMarker is a unique paragraph string inserted by expandBlankLines to
// represent extra blank lines that Goldmark would otherwise normalize away.
// It is removed from the rendered output by restoreExpandedBlanks.
const blankMarker = "GLANCEBLANK"

// preventSetextFromThematicBreaks inserts a blank line before any line that
// would otherwise combine with the preceding non-blank text line into a setext
// heading. This covers two cases:
//
//   - thematic breaks (≥3 of -, *, or _): "text\n---" should render as a
//     paragraph followed by a horizontal rule, not a setext H2.
//   - setext underlines (a run of only - or only =, any length): "text\n-" or
//     "text\n===" should stay a paragraph rather than turn the text above into
//     a heading. glance prefers explicit ATX headings (# / ##) so a stray
//     dash/equals line never silently rewrites the text above it.
func preventSetextFromThematicBreaks(src string) string {
	lines := strings.Split(src, "\n")
	out := make([]string, 0, len(lines)+4)
	for i, line := range lines {
		if i > 0 && len(out) > 0 && strings.TrimSpace(out[len(out)-1]) != "" &&
			(isThematicBreak(line) || isSetextUnderline(line)) {
			out = append(out, "")
		}
		out = append(out, line)
	}
	return strings.Join(out, "\n")
}

// isSetextUnderline reports whether line is a CommonMark setext heading
// underline: a non-empty run of only '-' or only '=' characters, with optional
// leading/trailing spaces but no interior spaces. A single character qualifies
// (CommonMark imposes no minimum length), so "-" and "=" both match.
func isSetextUnderline(line string) bool {
	trimmed := strings.TrimSpace(line)
	if trimmed == "" {
		return false
	}
	c := trimmed[0]
	if c != '-' && c != '=' {
		return false
	}
	for i := 0; i < len(trimmed); i++ {
		if trimmed[i] != c {
			return false
		}
	}
	return true
}

// isThematicBreak reports whether line is a CommonMark thematic break:
// at least 3 of the same character (-, *, or _) with optional interspersed spaces.
func isThematicBreak(line string) bool {
	trimmed := strings.TrimSpace(line)
	if len(trimmed) < 3 {
		return false
	}
	c := trimmed[0]
	if c != '-' && c != '*' && c != '_' {
		return false
	}
	count := 0
	for i := 0; i < len(trimmed); i++ {
		switch trimmed[i] {
		case c:
			count++
		case ' ', '\t':
			// spaces allowed between thematic-break chars
		default:
			return false
		}
	}
	return count >= 3
}

// expandBlankLines replaces runs of N≥2 consecutive blank lines between
// content blocks (outside fenced code blocks) with N-1 blankMarker paragraphs
// separated by single blank lines. Goldmark would otherwise collapse any run
// of blank lines to a single paragraph separator.
//
// Math: each inserted placeholder becomes its own paragraph in the rendered
// output, surrounded by Glamour's inter-paragraph blank line. Removing the
// placeholder lines (restoreExpandedBlanks) leaves exactly N blank lines in
// the final output.
func expandBlankLines(src string) string {
	lines := strings.Split(src, "\n")
	out := make([]string, 0, len(lines)+8)
	blanks := 0
	seenContent := false
	inFence := false

	flush := func() {
		if blanks == 0 {
			return
		}
		if seenContent && blanks >= 2 {
			out = append(out, "")
			for i := 1; i < blanks; i++ {
				out = append(out, blankMarker, "")
			}
		} else {
			for i := 0; i < blanks; i++ {
				out = append(out, "")
			}
		}
		blanks = 0
	}

	for _, line := range lines {
		trimmed := strings.TrimLeft(line, " \t")
		if strings.HasPrefix(trimmed, "```") || strings.HasPrefix(trimmed, "~~~") {
			flush()
			inFence = !inFence
			seenContent = true
			out = append(out, line)
			continue
		}
		if !inFence && strings.TrimSpace(line) == "" {
			blanks++
			continue
		}
		flush()
		seenContent = true
		out = append(out, line)
	}
	// Trailing blanks are preserved as-is (handled by countTrailingBlankLines).
	for i := 0; i < blanks; i++ {
		out = append(out, "")
	}
	return strings.Join(out, "\n")
}

// restoreExpandedBlanks removes blankMarker lines from the Glamour-rendered
// output, converting the surrounding blank lines back into the original
// multi-blank gap. Must be called before the leading/trailing blank restoration.
func restoreExpandedBlanks(rendered string) string {
	lines := strings.Split(rendered, "\n")
	out := lines[:0]
	for _, line := range lines {
		plain := strings.TrimSpace(ansiRE.ReplaceAllString(line, ""))
		if plain == blankMarker {
			continue
		}
		out = append(out, line)
	}
	return strings.Join(out, "\n")
}

// countLeadingBlankLines returns the number of blank lines at the start of s.
// A bare "\n" (just a line terminator) counts as 0; "\n\n" counts as 1.
func countLeadingBlankLines(s string) int {
	n := len(s) - len(strings.TrimLeft(s, "\n"))
	// For an all-blank document (no non-space content), treat the last \n as
	// the document terminator and subtract it from the count.
	if strings.TrimSpace(s) == "" {
		if n <= 1 {
			return 0
		}
		return n - 1
	}
	return n
}

// countTrailingBlankLines returns the number of blank lines at the end of s,
// not counting the line terminator of the last non-blank line.
// Examples:
//
//	"text\n"     → 0  (the trailing \n is just the line terminator)
//	"text\n\n"   → 1
//	"text\n\n\n" → 2
//	"\n"         → 0  (just a terminator, no blank-line content)
//	"\n\n"       → 1
func countTrailingBlankLines(s string) int {
	trimmed := strings.TrimRight(s, "\n")
	n := len(s) - len(trimmed) // number of trailing \n characters
	if n <= 1 {
		return 0
	}
	return n - 1
}

File: src/preprocess.c
/* preprocess.c — tolerant-Markdown source fix-ups (see preprocess.h).
 *
 * Works line by line. Bold tightening rewrites "** x **" to "**x**" outside of
 * code; setext neutralizing inserts a blank line before a stray "---"/"==="
 * underline so it can't absorb the paragraph above into a heading. Fenced and
 * inline code are left untouched.
 */
#include "preprocess.h"

#include <stdlib.h>
#include <string.h>

/* ---- tiny string builder -------------------------------------------------- */

typedef struct { char *data; size_t len, cap; } SB;

/* Append n bytes, growing the buffer as needed. */
static void sb_putn(SB *sb, const char *s, size_t n) {
    if (sb->len + n + 1 > sb->cap) {
        size_t cap = sb->cap ? sb->cap : 256;
        while (sb->len + n + 1 > cap) cap *= 2;
        char *p = realloc(sb->data, cap);
        if (!p) return;
        sb->data = p; sb->cap = cap;
    }
    memcpy(sb->data + sb->len, s, n);
    sb->len += n; sb->data[sb->len] = '\0';
}
static void sb_putc(SB *sb, char c) { sb_putn(sb, &c, 1); }

/* ---- line classification -------------------------------------------------- */

/* Offsets of the first and one-past-last non-space/tab byte of [s, s+n). */
static void trim_bounds(const char *s, size_t n, size_t *lo, size_t *hi) {
    size_t a = 0, b = n;
    while (a < b && (s[a] == ' ' || s[a] == '\t')) a++;
    while (b > a && (s[b - 1] == ' ' || s[b - 1] == '\t')) b--;
    *lo = a; *hi = b;
}

/* A fence marker line: trimmed text starts with ``` or ~~~. */
static int is_fence(const char *s, size_t n) {
    size_t lo, hi; trim_bounds(s, n, &lo, &hi);
    return hi - lo >= 3 &&
           ((s[lo] == '`' && s[lo+1] == '`' && s[lo+2] == '`') ||
            (s[lo] == '~' && s[lo+1] == '~' && s[lo+2] == '~'));
}

/* A setext underline: a run of only '-' or only '=' (length >= 1). */
static int is_setext_underline(const char *s, size_t n) {
    size_t lo, hi; trim_bounds(s, n, &lo, &hi);
    if (hi == lo) return 0;
    char c = s[lo];
    if (c != '-' && c != '=') return 0;
    for (size_t i = lo; i < hi; i++) if (s[i] != c) return 0;
    return 1;
}

/* A thematic break: >= 3 of one of - * _, with optional spaces between. */
static int is_thematic_break(const char *s, size_t n) {
    size_t lo, hi; trim_bounds(s, n, &lo, &hi);
    if (hi - lo < 3) return 0;
    char c = s[lo];
    if (c != '-' && c != '*' && c != '_') return 0;
    int count = 0;
    for (size_t i = lo; i < hi; i++) {
        if (s[i] == c) count++;
        else if (s[i] != ' ' && s[i] != '\t') return 0;
    }
    return count >= 3;
}

/* ---- bold tightening ------------------------------------------------------ */

/* Tighten one delimiter (d = '*' or '_', doubled) within a code-free segment:
 * "<d><d> x <d><d>" with no inner d becomes "<d><d>x<d><d>". Empty spans
 * ("** **") and already-tight spans are left as they are. */
static void tighten_segment(SB *out, const char *s, size_t n, char d) {
    size_t i = 0;
    while (i < n) {
        if (i + 1 < n && s[i] == d && s[i + 1] == d) {
            size_t k = i + 2;                 /* find the next d */
            while (k < n && s[k] != d) k++;
            if (k + 1 < n && s[k] == d && s[k + 1] == d) {
                size_t lo, hi;                /* inner = [i+2, k) */
                trim_bounds(s + i + 2, k - (i + 2), &lo, &hi);
                lo += i + 2; hi += i + 2;
                if (hi > lo && (lo != i + 2 || hi != k)) {
                    sb_putc(out, d); sb_putc(out, d);
                    sb_putn(out, s + lo, hi - lo);
                    sb_putc(out, d); sb_putc(out, d);
                } else {
                    sb_putn(out, s + i, k + 2 - i);   /* empty or already tight */
                }
                i = k + 2;
                continue;
            }
            sb_putn(out, s + i, 2);           /* unmatched: emit "<d><d>" */
            i += 2;
            continue;
        }
        sb_putc(out, s[i++]);
    }
}

/* Apply bold tightening to a line, copying inline code spans (`...`) verbatim. */
static void tighten_line(SB *out, const char *s, size_t n) {
    size_t i = 0;
    while (i < n) {
        if (s[i] == '`') {
            size_t j = i; while (j < n && s[j] == '`') j++;
            size_t run = j - i;
            size_t k = j, close = n;          /* find a backtick run of equal length */
            while (k < n) {
                if (s[k] == '`') {
                    size_t m = k; while (m < n && s[m] == '`') m++;
                    if (m - k == run) { close = k; break; }
                    k = m;
                } else k++;
            }
            if (close < n) { sb_putn(out, s + i, close + run - i); i = close + run; continue; }
            sb_putn(out, s + i, run); i = j; continue;   /* unterminated run */
        }
        size_t start = i;
        while (i < n && s[i] != '`') i++;
        /* tighten ** then __ over this code-free segment, via a scratch buffer */
        SB tmp = {0};
        tighten_segment(&tmp, s + start, i - start, '*');
        SB tmp2 = {0};
        tighten_segment(&tmp2, tmp.data ? tmp.data : "", tmp.len, '_');
        sb_putn(out, tmp2.data ? tmp2.data : "", tmp2.len);
        free(tmp.data); free(tmp2.data);
    }
}

/* ---- driver --------------------------------------------------------------- */

char *preprocess(const char *src, size_t len, size_t *out_len) {
    SB out = {0};
    int in_fence = 0, prev_nonblank = 0;
    size_t i = 0;
    while (i <= len) {
        size_t start = i;
        while (i < len && src[i] != '\n') i++;
        size_t llen = i - start;             /* current line, sans newline */
        const char *line = src + start;

        int fence = is_fence(line, llen);
        if (!in_fence && !fence && prev_nonblank &&
            (is_thematic_break(line, llen) || is_setext_underline(line, llen)))
            sb_putc(&out, '\n');             /* keep it off the paragraph above */

        if (fence) {
            in_fence = !in_fence;
            sb_putn(&out, line, llen);
        } else if (in_fence) {
            sb_putn(&out, line, llen);       /* code content: verbatim */
        } else {
            tighten_line(&out, line, llen);
        }

        size_t lo, hi; trim_bounds(line, llen, &lo, &hi);
        prev_nonblank = hi > lo;

        if (i < len) sb_putc(&out, '\n');    /* re-add the separator */
        i++;                                  /* step past the '\n' (or end) */
    }
    if (!out.data) out.data = calloc(1, 1);
    if (out_len) *out_len = out.len;
    return out.data;
}
